client/src/util/exponentialBackoffFetch.js

```diff
@@ -0,0 +1,68 @@
+export default function createBackoffFetch(retryTimes = 3) {
+  let prevHash = null;
+  let retryCount = 0;
+  let response = null;
+  let error = null;
+  let promise = null;
+
+  // This will only be called by itself or when we've seen a new matrix
+  const backoffFetch = (fetchArgs) => {
+    const promiseExecuter = (resolve, reject) => {
+      retryCount += 1;
+      // We don't want to timeout our first request
+      if (retryCount === 1) {
+        fetch(...fetchArgs)
+          .then((resp) => {
+            response = resp;
+            if (resp.ok) {
+              resolve(response);
+            } else {
+              promiseExecuter(resolve, reject);
+            }
+          })
+          .catch((e) => {
+            response = null;
+            error = e;
+            promiseExecuter(resolve, reject);
+          });
+      } else {
+        // Timeout all other requests
+        setTimeout(async () => {
+          try {
+            const resp = await fetch(...fetchArgs);
+            response = resp;
+            // Fetch status 2xx
+            if (resp.ok) {
+              resolve(response);
+            } else if (retryCount - 1 === retryTimes) {
+              if (error) reject(error);
+              resolve(response);
+            } else {
+              // Some non 2xx response code
+              promiseExecuter(resolve, reject);
+            }
+          } catch (e) {
+            // Fetch API error
+            error = e;
+            if (retryCount - 1 === retryTimes) reject(error);
+            promiseExecuter(resolve, reject);
+          }
+          // delay amount (3.5, 17.5, 87.5, 437.5 seconds)
+        }, 700 * 5 ** (retryCount - 1));
+      }
+    };
+    return new Promise(promiseExecuter);
```

Comment:
since `async () => {}` already returns a promise, maybe we don't need `new Promise()`?

either `return response` or `throw error` should resolve/reject the async function's promise!

Reply:
We need to inspect the response before we hand it back to the user. If response is not OK, we need to fetch again.  In the fetch API a response is a fulfilled promise regardless of the response code, but not to our expBackoffFetch API (non 2xx code only fulfills the promise once we expended all attemps)

Reply:
Ahh yeah, you're right! We can't avoid `new Promise`, since we need to inspect `response.ok` and `fetch()` happens in `setTimeout()`

PTAL this PR! https://github.com/chanzuckerberg/cellxgene/pull/2003#pullrequestreview-551089371

Thank you!